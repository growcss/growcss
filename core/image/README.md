# GrowCss Image

Speedy, optimized images without the work.

 It provides the fastest, most optimized image loading performance possible for React websites.

## Out of the box it

- Loads the optimal size of image for each device size and screen resolution
- Holds the image position while loading so your page doesn’t jump around as images load
- Uses the “blur-up” effect i.e. it loads a tiny version of the image to show while the full image is loading
- Alternatively provides a “traced placeholder” SVG of the image.
- Lazy loads images which reduces bandwidth and speeds the initial load time
- Uses WebP images if browser supports the format

## Installation

```bash
npm install @growcss/image
```

or if you're using yarn

```bash
yarn add @growcss/image
```

## Usage

To implement Image component into your project you'll need to add the import:

```js
import {Image} from '@growcss/image';
```

After adding import into your project you can use it simply like:

```jsx
<Image alt="your alt text" src="your image" />
```

## GrowCss Image srcSet

The `srcSet` attribute is a string consisting of one or more image selection strings separated by commas. 
Each image selection string is composed of:

1. A url to an image
2. One or more spaces as a separator
3. One or _more_ descriptors separated by spaces
    - width descriptor: a positive integer directly followed by 'w'. e.g. `700w`
    - height descriptor: a positive integer directly followed by 'h'. e.g. `345h`
    - pixel density descriptor: a positive floating point number directly followed by 'x'. e.g. `2.0x`
    - media descriptor: used for the media type in source element

`GrowCss image` extends the `<Image srcSet="...">` feature by allowing multiple descriptors for an image
and mixed descriptors in a single srcSet.

### Example image srcSet

`srcSet="foo-s.jpg 150w, foo-sh.jpg 150w 2.0x, foo-m.jpg 405w, foo-mh 2.0x 405w, foo-l 1024w, foo-t 500w 750h media=(min-width: 300px)"`

### Automatic Density

If _none_ of the image selection strings includes a pixel density descriptor ('x' e.g. `4.0x`), then the image
selection process will automatically compensate for the viewport's pixel density.
