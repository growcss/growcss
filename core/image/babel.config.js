module.exports = require('./node_modules/@build/extended-babel/babel.config.js');
