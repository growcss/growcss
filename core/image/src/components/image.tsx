import React, { Component, ReactNode, ReactElement } from 'react';
// eslint-disable-next-line no-unused-vars
import { withTheme, DefaultTheme } from 'styled-components';
// eslint-disable-next-line no-unused-vars
import { getThemeValue } from '@growcss/theme';
import { stripUnit } from '@growcss/elaborate';
import { AspectRatioPlaceholder } from '../styled/aspect-ratio-placeholder';
import { FigureElement } from '../styled/figure-element';
import { PlaceholderElement } from '../styled/placeholder-element';
import { ImageStateType } from '../states';
import { ImageProps, CandidateProps, PictureProps } from '../../types';
import { getDimension } from './helper/get-dimension';
import { parseSrcSet } from './helper/srcset-parser';
import { filterSmall } from './helper/filter-small';
import { filterLarge } from './helper/filter-large';
import { isSsr } from './helper/is-ssr';
import { isNativeConnection } from './helper/is-native-connection';
import { Picture } from './picture';

interface DefaultImageProps {
  onLoad: () => void;
  onError: () => void;
  onStartLoad: () => void;
  onOffline: () => void;
  onSlowNetwork: () => void;
  preload: boolean;
  lazy: boolean;
  webpRegex: RegExp;
  observer: {
    rootMargin: string;
    threshold: number | number[];
  };
}

type PropsWithDefaults = ImageProps & DefaultImageProps;

// Cache if we've seen an image before so we don't bother with
// lazy-loading & fading in on subsequent mounts.
const imageCache: { [key: string]: boolean } = {};

const inImageCache = (src: string): boolean => {
  return imageCache[src] || false;
};

const activateCacheForImage = (src: string): void => {
  imageCache[src] = true;
};

const defaultObserver = {
  rootMargin: '0px 0px 0px 0px',
  threshold: 0,
};

class ExtendedImage extends Component<ImageProps, ImageStateType> {
  public static defaultProps: DefaultImageProps = {
    onLoad: () => {},
    onError: () => {},
    onStartLoad: () => {},
    onOffline: () => {},
    onSlowNetwork: () => {},
    preload: false,
    lazy: false,
    webpRegex: new RegExp('/.+.webp$/i'),
    observer: defaultObserver,
  };

  /**
   * @param {PropsWithDefaults} props
   */
  public constructor(props: PropsWithDefaults) {
    super(props);

    let inViewport = true;

    const {
      src,
      srcSet,
      placeholder,
      theme,
      webpRegex,
      lazy,
    } = props as PropsWithDefaults;
    const srcArray =
      srcSet !== undefined
        ? parseSrcSet(srcSet, getThemeValue('image.breakpoints', {})(theme))
        : [];
    const hasWebpImages =
      srcArray.filter(srcItem => {
        return webpRegex.test(srcItem.url);
      }).length > 0;
    const selectedImage = ExtendedImage.bestMatchingImage(srcArray, src);
    const seenBefore = inImageCache(selectedImage);
    const hasSrcSet = srcArray.length > 1;

    // browser with Intersection Observer available
    if (
      !seenBefore &&
      lazy &&
      typeof window !== `undefined` &&
      window.IntersectionObserver
    ) {
      inViewport = false;
    }

    // Never render image during SSR
    if (typeof window === `undefined` && lazy) {
      inViewport = false;
    }

    this.state = {
      loadState: 'initial',
      possiblySlowNetwork: false,
      inViewport,
      online: true,
      hasWebp: hasWebpImages,
      srcSetWebpArray:
        hasWebpImages && hasSrcSet
          ? ExtendedImage.filterWebp(webpRegex, srcArray, true)
          : undefined,
      srcSetArray: hasSrcSet
        ? ExtendedImage.filterWebp(webpRegex, srcArray, false)
        : undefined,
      placeholder: ExtendedImage.getPlaceholder(srcArray, placeholder),
      src: selectedImage,
      seenBefore,
    };

    this.updateConnection = this.updateConnection.bind(this);
    this.possiblySlowNetworkListener = this.possiblySlowNetworkListener.bind(this);
    this.updateOnlineStatus = this.updateOnlineStatus.bind(this);
    this.onEnter = this.onEnter.bind(this);
  }

  /**
   * {@inheritdoc}
   */
  public componentDidMount(): void {
    const { threshold, onOffline, onSlowNetwork } = this.props as PropsWithDefaults;
    const isOnline = navigator.onLine;

    if (isNativeConnection && isOnline) {
      navigator.connection.addEventListener('onchange', this.updateConnection);
    } else if (threshold !== undefined) {
      window.document.addEventListener(
        'possiblySlowNetwork',
        this.possiblySlowNetworkListener,
      );
    }

    window.addEventListener('online', this.updateOnlineStatus);
    window.addEventListener('offline', this.updateOnlineStatus);

    const { online, possiblySlowNetwork, src } = this.state;

    if (!online && typeof onOffline === 'function') {
      onOffline();
    }

    if (possiblySlowNetwork && typeof onSlowNetwork === 'function') {
      onSlowNetwork();
    }

    activateCacheForImage(src);
  }

  /**
   * {@inheritdoc}
   */
  public componentWillUnmount(): void {
    const { threshold } = this.props;

    if (isNativeConnection) {
      navigator.connection.removeEventListener('onchange', this.updateConnection);
    } else if (threshold !== undefined) {
      window.document.removeEventListener(
        'possiblySlowNetwork',
        this.possiblySlowNetworkListener,
      );
    }

    window.removeEventListener('online', this.updateOnlineStatus);
    window.removeEventListener('offline', this.updateOnlineStatus);
  }

  public onEnter(event, unobserve): void {
    const { inViewport } = this.state;

    if (inViewport) {
      return;
    }

    const { lazy } = this.props;

    if (event.isIntersecting || lazy === false) {
      unobserve();
    }

    this.setState({
      inViewport: !lazy ? true : event.isIntersecting,
    });
  }

  /**
   * {@inheritdoc}
   */
  public render(): ReactNode {
    const { observer, lazy } = this.props as PropsWithDefaults;

    if (lazy) {
      try {
        // eslint-disable-next-line global-require
        const IntersectionObserver = require('@researchgate/react-intersection-observer')
          .default;

        return (
          <IntersectionObserver
            onChange={this.onEnter}
            rootMargin={observer.rootMargin || defaultObserver.rootMargin}
            threshold={observer.threshold || defaultObserver.threshold}
          >
            {this.renderImage()}
          </IntersectionObserver>
        );
      } catch (error) {
        if (error instanceof Error && error.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            'Please run "npm install @researchgate/react-intersection-observer@^1.0.0-beta.0" to use the "lazy" feature of the growcss image component.',
          );
        }
      }
    }

    return this.renderImage();
  }

  /**
   * Update the connection state.
   *
   * @returns {void}
   */
  public updateConnection(): void {
    if (!navigator.onLine) {
      return;
    }

    const { loadState } = this.state;

    if (loadState === 'initial') {
      // @todo use this for a download icon
      /* eslint-disable react/no-unused-state */
      this.setState({
        connection: {
          effectiveType: navigator.connection.effectiveType,
          downlink: navigator.connection.downlink,
          rtt: navigator.connection.rtt,
        },
      });
    }
  }

  /**
   * Update the online state.
   *
   * @returns {void}
   */
  public updateOnlineStatus(): void {
    this.setState({ online: navigator.onLine });
  }

  /**
   * Check if the network is slow and update state.
   *
   * @param {object} erorr
   *
   * @returns void
   */
  public possiblySlowNetworkListener(erorr): void {
    const { loadState, possiblySlowNetwork: statePossiblySlowNetwork } = this.state;

    if (loadState !== 'initial') {
      return;
    }

    const { possiblySlowNetwork } = erorr.detail;

    if (!statePossiblySlowNetwork && possiblySlowNetwork) {
      this.setState({ possiblySlowNetwork });
    }
  }

  /**
   * Render a default placeholder if no placeholder was found.
   *
   * @param {DefaultTheme} theme
   * @param {undefined | string} alt
   * @param {undefined | string} placeholder
   *
   * @returns {ReactNode}
   */
  public static renderPlaceholder(
    theme: DefaultTheme,
    alt?: string,
    placeholder?: string,
  ): ReactNode {
    const defaultSvgPlaceholder = `<svg xmlns='http://www.w3.org/2000/svg' width='1' height='1'><rect width='1' height='1' fill='${getThemeValue(
      'image.backgroundColor',
    )(theme)}'/></svg>`;

    return (
      <PlaceholderElement
        className="placeholder"
        src={placeholder || `data:image/svg+xml;base64,${btoa(defaultSvgPlaceholder)}`}
        crossOrigin="anonymous"
        alt={alt}
      />
    );
  }

  /**
   * @param {string | number | undefined} height
   * @param {string | number | undefined} width
   */
  public static getDimensionRatio(
    height: string | number | undefined,
    width: string | number | undefined,
  ): number {
    if (height !== undefined && width !== undefined) {
      return (stripUnit(height) / stripUnit(width)) * 100;
    }

    const { height: dHeight, width: dWidth } = getDimension();

    return (dHeight / dWidth) * 100;
  }

  /**
   * Get best image from srcset
   *
   * @param {CandidateProps[]} srcSetArray
   * @param {undefined|string} src
   *
   * @returns {string}
   */
  public static bestMatchingImage(srcSetArray: CandidateProps[], src?: string): string {
    let densityMultiplier = 1;

    // Check if denisty is provided in srcset
    let hasDensity = false;

    srcSetArray.forEach(srcSet => {
      if ('density' in srcSet) {
        hasDensity = true;
      }
    });

    if (!hasDensity) {
      // When no density is set use devicePixelRatio as a multiplier for reference dimensions
      densityMultiplier = window.devicePixelRatio || 1.0;
    }

    const refDim: { width: number; height: number } = getDimension();
    let filtered: CandidateProps[] = srcSetArray;

    if (typeof src === 'string' && src !== '') {
      filtered = srcSetArray.concat([
        {
          url: src,
        },
      ]);
    }

    filtered = filterLarge(
      filterLarge(
        filterLarge(filtered, 'width', refDim.width * densityMultiplier),
        'height',
        refDim.height * densityMultiplier,
      ),
      'density',
      !hasDensity ? 1 : window.devicePixelRatio || 1.0, // When no density is provided densityMultiplier is used and density filter can be set to 1, otherwise use devicePixelRatio
    );

    filtered = filterSmall(
      filterSmall(filterSmall(filtered, 'width'), 'height'),
      'density',
    );

    return filtered[0].url;
  }

  /**
   * Returns a filtered copy of the source set array
   * including only webp images if supports is true,
   * or only non webp images if supports is false.
   *
   * @param {RegExp}           webpRegex
   * @param {CandidateProps[]} srcArray
   * @param {boolean}          supports Indicates if the browser supports webp.
   *
   * @returns CandidateProps[] filtered source set array.
   */
  public static filterWebp(
    webpRegex: RegExp,
    srcArray: CandidateProps[],
    supports: boolean,
  ): CandidateProps[] {
    return srcArray.filter(srcItem => {
      return webpRegex.test(srcItem.url) === supports;
    });
  }

  /**
   *
   * @param {CandidateProps[]} srcArray
   * @param {undefined | string} placeholder
   */
  public static getPlaceholder(
    srcArray: CandidateProps[],
    placeholder?: string,
  ): undefined | string {
    if (placeholder === undefined && srcArray.length !== 0) {
      const smallImages = filterSmall(
        filterSmall(filterSmall(srcArray, 'width'), 'height'),
        'density',
      );

      return smallImages.length !== 0 ? smallImages[0].url : undefined;
    }

    return placeholder;
  }

  /**
   * Render Image component.
   *
   * @returns {ReactElement<PictureProps>}
   */
  private renderImage(): ReactElement<PictureProps> {
    const {
      children,
      height,
      width,
      preload,
      theme,
      onError,
      onLoad,
      onStartLoad,
      alt,
      sizes,
      crossOrigin,
      ...other
    } = this.props as PropsWithDefaults;

    const {
      src,
      srcSetWebpArray,
      srcSetArray,
      placeholder,
      hasWebp,
      loadState,
      inViewport,
      seenBefore,
    } = this.state;

    const fadeIn = preload && !seenBefore;

    return (
      <FigureElement className="gc-image">
        <AspectRatioPlaceholder>
          <div
            style={{
              width: '100%',
              paddingBottom: `${ExtendedImage.getDimensionRatio(height, width)}%`,
            }}
          />
          {fadeIn && ExtendedImage.renderPlaceholder(theme, alt, placeholder)}
          {inViewport && (
            <Picture
              src={src}
              state={loadState}
              preload={preload}
              srcSetArray={srcSetArray}
              srcSetWebpArray={srcSetWebpArray}
              hasWebp={hasWebp}
              onLoad={onLoad}
              onError={onError}
              onStartLoad={onStartLoad}
              crossOrigin={crossOrigin}
              sizes={sizes}
              alt={alt}
              {...other}
            />
          )}
          {isSsr && (
            <noscript>
              <Picture
                src={src}
                state="loaded"
                preload={false}
                srcSetArray={srcSetArray}
                srcSetWebpArray={srcSetWebpArray}
                hasWebp
                crossOrigin={crossOrigin}
                sizes={sizes}
                alt={alt}
                {...other}
              />
            </noscript>
          )}
        </AspectRatioPlaceholder>
        {children}
      </FigureElement>
    );
  }
}

export const StyledImage = withTheme(ExtendedImage);
