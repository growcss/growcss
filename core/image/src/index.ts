export { StyledImage as Image } from './components/image';
export { FigcaptionElement as Figcaption } from './styled/figcaption-element';
