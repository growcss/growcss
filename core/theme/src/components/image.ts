import { colors } from '../css/colors';

export default {
  breakpoints: {
    medium: '640w',
    large: '1024w',
    xlarge: '1200w',
    xxlarge: '1440w',
  },
  previewBackgroundColor: colors.greyLight,
};
