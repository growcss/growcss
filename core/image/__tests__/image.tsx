import * as React from 'react';
import { mount } from 'enzyme';
import { axe, toHaveNoViolations } from 'jest-axe';
import { GrowCss } from '@growcss/theme';
import { StyledImage as Image } from '../src/components/image';

window.IntersectionObserver.prototype.USE_MUTATION_OBSERVER = false;

expect.extend(toHaveNoViolations);

test('Image load small image if no placeholder image is set', () => {
  const wrapper = mount(
    <Image
      srcSet="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg medium"
      preload
      theme={GrowCss}
    />,
  );

  expect(wrapper).toMatchSnapshot();

  const placeholder = wrapper.find('.gc-image img.placeholder').html();

  expect(placeholder).toContain('img class="placeholder');
  expect(placeholder).toContain(
    'src="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg"',
  );
  expect(placeholder).toContain('crossorigin="anonymous"');
});

test('Image a11y support', async () => {
  const wrapper = mount(
    <Image
      srcSet="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg"
      alt="test image"
      theme={GrowCss}
    />,
  );

  expect(await axe(wrapper.html())).toHaveNoViolations();
});

test(`should have correct src, title and alt attributes`, () => {
  const wrapper = mount(
    <Image
      srcSet="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg medium, https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/600x/mountain.jpg large"
      title="Title for the image"
      alt="Alt text for the image"
      theme={GrowCss}
    />,
  );

  expect(wrapper).toMatchSnapshot();

  const pictureImage = wrapper.find('.gc-image picture img').html();
  const sourcePictureImage = wrapper.find('.gc-image picture source').html();

  expect(pictureImage).toContain(
    `src="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/600x/mountain.jpg"`,
  );
  expect(sourcePictureImage).toContain(
    `srcset="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg 640w, https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/600x/mountain.jpg 1024w"`,
  );
  expect(pictureImage).toContain(`title="Title for the image"`);
  expect(pictureImage).toContain(`alt="Alt text for the image"`);
});
