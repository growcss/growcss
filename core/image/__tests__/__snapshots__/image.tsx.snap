// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image load small image if no placeholder image is set 1`] = `
<WithTheme(ExtendedImage)
  preload={true}
  srcSet="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg medium"
  theme={
    Object {
      "body": Object {
        "backgroundColor": "#ffffff",
      },
      "colors": Object {
        "black": "#22292f",
        "blue": "#3490dc",
        "blueDark": "#2779bd",
        "blueDarker": "#1c3d5a",
        "blueDarkest": "#12283a",
        "blueLight": "#6cb2eb",
        "blueLighter": "#bcdefa",
        "blueLightest": "#eff8ff",
        "green": "#38c172",
        "greenDark": "#1f9d55",
        "greenDarker": "#1a4731",
        "greenDarkest": "#0f2f21",
        "greenLight": "#51d88a",
        "greenLighter": "#a2f5bf",
        "greenLightest": "#e3fcec",
        "grey": "#b8c2cc",
        "greyDark": "#8795a1",
        "greyDarker": "#606f7b",
        "greyDarkest": "#3d4852",
        "greyLight": "#dae1e7",
        "greyLighter": "#f1f5f8",
        "greyLightest": "#f8fafc",
        "indigo": "#6574cd",
        "indigoDark": "#5661b3",
        "indigoDarker": "#2f365f",
        "indigoDarkest": "#191e38",
        "indigoLight": "#7886d7",
        "indigoLighter": "#b2b7ff",
        "indigoLightest": "#e6e8ff",
        "orange": "#f6993f",
        "orangeDark": "#de751f",
        "orangeDarker": "#613b1f",
        "orangeDarkest": "#462a16",
        "orangeLight": "#faad63",
        "orangeLighter": "#fcd9b6",
        "orangeLightest": "#fff5eb",
        "pink": "#f66d9b",
        "pinkDark": "#eb5286",
        "pinkDarker": "#6f213f",
        "pinkDarkest": "#451225",
        "pinkLight": "#fa7ea8",
        "pinkLighter": "#ffbbca",
        "pinkLightest": "#ffebef",
        "purple": "#9561e2",
        "purpleDark": "#794acf",
        "purpleDarker": "#382b5f",
        "purpleDarkest": "#21183c",
        "purpleLight": "#a779e9",
        "purpleLighter": "#d6bbfc",
        "purpleLightest": "#f3ebff",
        "red": "#e3342f",
        "redDark": "#cc1f1a",
        "redDarker": "#621b18",
        "redDarkest": "#3b0d0c",
        "redLight": "#ef5753",
        "redLighter": "#f9acaa",
        "redLightest": "#fcebea",
        "teal": "#4dc0b5",
        "tealDark": "#38a89d",
        "tealDarker": "#20504f",
        "tealDarkest": "#0d3331",
        "tealLight": "#64d5ca",
        "tealLighter": "#a0f0ed",
        "tealLightest": "#e8fffe",
        "transparent": "transparent",
        "white": "#ffffff",
        "yellow": "#ffed4a",
        "yellowDark": "#f2d024",
        "yellowDarker": "#684f1d",
        "yellowDarkest": "#453411",
        "yellowLight": "#fff382",
        "yellowLighter": "#fff9c2",
        "yellowLightest": "#fcfbeb",
      },
      "global": Object {
        "breakpoints": Object {
          "large": 1024,
          "medium": 640,
          "small": 0,
          "xlarge": 1200,
          "xxlarge": 1440,
        },
        "hidpiBreakpoints": Object {
          "hidpi-1": 1,
          "hidpi-1-5": 1.5,
          "hidpi-2": 2,
          "hidpi-3": 3,
          "retina": 2,
        },
        "margin": "1rem",
        "padding": "1rem",
        "position": "1rem",
        "printBreakpoint": "large",
        "stdWebDpi": 96,
        "textDirection": "ltr",
      },
      "grid": Object {
        "columns": 12,
        "gutterBreakpoint": "small",
        "marginGutters": Object {
          "medium": "30px",
          "small": "20px",
        },
        "maxWidth": "75rem",
        "paddingGutters": Object {
          "medium": "30px",
          "small": "20px",
        },
      },
      "image": Object {
        "breakpoints": Object {
          "large": "1024w",
          "medium": "640w",
          "xlarge": "1200w",
          "xxlarge": "1440w",
        },
        "previewBackgroundColor": "#dae1e7",
      },
      "shadows": Object {
        "default": "0 2px 4px 0 rgba(0,0,0,0.10)",
        "inner": "inset 0 2px 4px 0 rgba(0,0,0,0.06)",
        "lg": "0 15px 30px 0 rgba(0,0,0,0.11), 0 5px 15px 0 rgba(0,0,0,0.08)",
        "md": "0 4px 8px 0 rgba(0,0,0,0.12), 0 2px 4px 0 rgba(0,0,0,0.08)",
        "none": "none",
        "outline": "0 0 0 3px rgba(52,144,220,0.5)",
      },
      "spinner": Object {
        "color": "#22292f",
        "size": Object {
          "large": 24,
          "medium": 16,
          "small": 8,
          "xlarge": 48,
          "xxlarge": 96,
        },
      },
      "typography": Object {
        "antialiased": true,
        "breakpoints": Object {
          "medium": Array [
            3.052,
            2.441,
            1.953,
            1.563,
            1.25,
            1,
          ],
          "small": Array [
            2.488,
            2.074,
            1.728,
            1.44,
            1.2,
            1,
          ],
        },
        "codeFontFamily": "\\"SFMono-Medium\\", \\"SF Mono\\", \\"Segoe UI Mono\\", \\"Roboto Mono\\", \\"Ubuntu Mono\\", Menlo, Consolas, Courier, monospace",
        "color": "#22292f",
        "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
        "fontSize": "16px",
        "lineHeight": 1.45,
        "weightBold": "bold",
        "weightNormal": 600,
      },
      "zIndex": Object {
        "0": 0,
        "10": 10,
        "20": 20,
        "30": 30,
        "40": 40,
        "50": 50,
        "auto": "auto",
      },
    }
  }
>
  <figure-element__FigureElement
    className="gc-image"
  >
    <figure
      className="gc-image figure-element__FigureElement-sc-92c2gs-0 gOCodI"
    >
      <aspect-ratio-placeholder__AspectRatioPlaceholder>
        <div
          className="aspect-ratio-placeholder__AspectRatioPlaceholder-sc-1p97xwn-0 bhPjyK"
        >
          <div
            style={
              Object {
                "paddingBottom": "75%",
                "width": "100%",
              }
            }
          />
          <placeholder-element__PlaceholderElement
            className="placeholder"
            crossOrigin="anonymous"
            src="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg"
          >
            <img
              className="placeholder placeholder-element__PlaceholderElement-sc-1fnxcwp-0 fQZFih"
              crossOrigin="anonymous"
              src="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg"
            />
          </placeholder-element__PlaceholderElement>
          <picture-element__PictureElement
            className="preload"
          >
            <picture
              className="preload picture-element__PictureElement-sc-5wdao-0 jPicbX"
            >
              <img
                onError={[Function]}
                onLoad={[Function]}
                src="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg"
              />
            </picture>
          </picture-element__PictureElement>
        </div>
      </aspect-ratio-placeholder__AspectRatioPlaceholder>
    </figure>
  </figure-element__FigureElement>
</WithTheme(ExtendedImage)>
`;

exports[`should have correct src, title and alt attributes 1`] = `
<WithTheme(ExtendedImage)
  alt="Alt text for the image"
  srcSet="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg medium, https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/600x/mountain.jpg large"
  theme={
    Object {
      "body": Object {
        "backgroundColor": "#ffffff",
      },
      "colors": Object {
        "black": "#22292f",
        "blue": "#3490dc",
        "blueDark": "#2779bd",
        "blueDarker": "#1c3d5a",
        "blueDarkest": "#12283a",
        "blueLight": "#6cb2eb",
        "blueLighter": "#bcdefa",
        "blueLightest": "#eff8ff",
        "green": "#38c172",
        "greenDark": "#1f9d55",
        "greenDarker": "#1a4731",
        "greenDarkest": "#0f2f21",
        "greenLight": "#51d88a",
        "greenLighter": "#a2f5bf",
        "greenLightest": "#e3fcec",
        "grey": "#b8c2cc",
        "greyDark": "#8795a1",
        "greyDarker": "#606f7b",
        "greyDarkest": "#3d4852",
        "greyLight": "#dae1e7",
        "greyLighter": "#f1f5f8",
        "greyLightest": "#f8fafc",
        "indigo": "#6574cd",
        "indigoDark": "#5661b3",
        "indigoDarker": "#2f365f",
        "indigoDarkest": "#191e38",
        "indigoLight": "#7886d7",
        "indigoLighter": "#b2b7ff",
        "indigoLightest": "#e6e8ff",
        "orange": "#f6993f",
        "orangeDark": "#de751f",
        "orangeDarker": "#613b1f",
        "orangeDarkest": "#462a16",
        "orangeLight": "#faad63",
        "orangeLighter": "#fcd9b6",
        "orangeLightest": "#fff5eb",
        "pink": "#f66d9b",
        "pinkDark": "#eb5286",
        "pinkDarker": "#6f213f",
        "pinkDarkest": "#451225",
        "pinkLight": "#fa7ea8",
        "pinkLighter": "#ffbbca",
        "pinkLightest": "#ffebef",
        "purple": "#9561e2",
        "purpleDark": "#794acf",
        "purpleDarker": "#382b5f",
        "purpleDarkest": "#21183c",
        "purpleLight": "#a779e9",
        "purpleLighter": "#d6bbfc",
        "purpleLightest": "#f3ebff",
        "red": "#e3342f",
        "redDark": "#cc1f1a",
        "redDarker": "#621b18",
        "redDarkest": "#3b0d0c",
        "redLight": "#ef5753",
        "redLighter": "#f9acaa",
        "redLightest": "#fcebea",
        "teal": "#4dc0b5",
        "tealDark": "#38a89d",
        "tealDarker": "#20504f",
        "tealDarkest": "#0d3331",
        "tealLight": "#64d5ca",
        "tealLighter": "#a0f0ed",
        "tealLightest": "#e8fffe",
        "transparent": "transparent",
        "white": "#ffffff",
        "yellow": "#ffed4a",
        "yellowDark": "#f2d024",
        "yellowDarker": "#684f1d",
        "yellowDarkest": "#453411",
        "yellowLight": "#fff382",
        "yellowLighter": "#fff9c2",
        "yellowLightest": "#fcfbeb",
      },
      "global": Object {
        "breakpoints": Object {
          "large": 1024,
          "medium": 640,
          "small": 0,
          "xlarge": 1200,
          "xxlarge": 1440,
        },
        "hidpiBreakpoints": Object {
          "hidpi-1": 1,
          "hidpi-1-5": 1.5,
          "hidpi-2": 2,
          "hidpi-3": 3,
          "retina": 2,
        },
        "margin": "1rem",
        "padding": "1rem",
        "position": "1rem",
        "printBreakpoint": "large",
        "stdWebDpi": 96,
        "textDirection": "ltr",
      },
      "grid": Object {
        "columns": 12,
        "gutterBreakpoint": "small",
        "marginGutters": Object {
          "medium": "30px",
          "small": "20px",
        },
        "maxWidth": "75rem",
        "paddingGutters": Object {
          "medium": "30px",
          "small": "20px",
        },
      },
      "image": Object {
        "breakpoints": Object {
          "large": "1024w",
          "medium": "640w",
          "xlarge": "1200w",
          "xxlarge": "1440w",
        },
        "previewBackgroundColor": "#dae1e7",
      },
      "shadows": Object {
        "default": "0 2px 4px 0 rgba(0,0,0,0.10)",
        "inner": "inset 0 2px 4px 0 rgba(0,0,0,0.06)",
        "lg": "0 15px 30px 0 rgba(0,0,0,0.11), 0 5px 15px 0 rgba(0,0,0,0.08)",
        "md": "0 4px 8px 0 rgba(0,0,0,0.12), 0 2px 4px 0 rgba(0,0,0,0.08)",
        "none": "none",
        "outline": "0 0 0 3px rgba(52,144,220,0.5)",
      },
      "spinner": Object {
        "color": "#22292f",
        "size": Object {
          "large": 24,
          "medium": 16,
          "small": 8,
          "xlarge": 48,
          "xxlarge": 96,
        },
      },
      "typography": Object {
        "antialiased": true,
        "breakpoints": Object {
          "medium": Array [
            3.052,
            2.441,
            1.953,
            1.563,
            1.25,
            1,
          ],
          "small": Array [
            2.488,
            2.074,
            1.728,
            1.44,
            1.2,
            1,
          ],
        },
        "codeFontFamily": "\\"SFMono-Medium\\", \\"SF Mono\\", \\"Segoe UI Mono\\", \\"Roboto Mono\\", \\"Ubuntu Mono\\", Menlo, Consolas, Courier, monospace",
        "color": "#22292f",
        "fontFamily": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
        "fontSize": "16px",
        "lineHeight": 1.45,
        "weightBold": "bold",
        "weightNormal": 600,
      },
      "zIndex": Object {
        "0": 0,
        "10": 10,
        "20": 20,
        "30": 30,
        "40": 40,
        "50": 50,
        "auto": "auto",
      },
    }
  }
  title="Title for the image"
>
  <figure-element__FigureElement
    className="gc-image"
  >
    <figure
      className="gc-image figure-element__FigureElement-sc-92c2gs-0 gOCodI"
    >
      <aspect-ratio-placeholder__AspectRatioPlaceholder>
        <div
          className="aspect-ratio-placeholder__AspectRatioPlaceholder-sc-1p97xwn-0 bhPjyK"
        >
          <div
            style={
              Object {
                "paddingBottom": "75%",
                "width": "100%",
              }
            }
          />
          <picture-element__PictureElement
            className="visible"
          >
            <picture
              className="visible picture-element__PictureElement-sc-5wdao-0 jPicbX"
            >
              <source
                key="-1954652935"
                srcSet="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/400x/mountain.jpg 640w, https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/600x/mountain.jpg 1024w"
              />
              <img
                alt="Alt text for the image"
                onError={[Function]}
                onLoad={[Function]}
                src="https://ucarecdn.com/1b73ebf8-b2d6-40cc-abd6-945d4df883c9/-/resize/600x/mountain.jpg"
                title="Title for the image"
              />
            </picture>
          </picture-element__PictureElement>
        </div>
      </aspect-ratio-placeholder__AspectRatioPlaceholder>
    </figure>
  </figure-element__FigureElement>
</WithTheme(ExtendedImage)>
`;
